contracts/koh/contracts/KingAutomaton.sol

```diff
@@ -92,28 +92,165 @@ contract KingAutomaton {
   // Treasury
   //////////////////////////////////////////////////////////////////////////////////////////////////
 
-  enum State {
-    Proposed,
-    Approved,
-    Rejected,
-    Accepted,
-    Terminated,
-    Completed
-  }
+  int256 constant INT256_MIN = int256(uint256(1) << 255);
+  int256 constant INT256_MAX = int256(~(uint256(1) << 255));
+  uint256 constant UINT256_MIN = 0;
+  uint256 constant UINT256_MAX = ~uint256(0);
+  uint256 constant MSB_SET = 1 << 255;
+  uint256 constant ALL_BUT_MSB = MSB_SET - 1;
+
+  uint256 constant NUM_CHOICES = 2;
+  uint256 constant BITS_PER_VOTE = 2;
+  uint256 constant VOTES_PER_WORD = 127;
+  uint256 constant BASE_MASK = (1 << BITS_PER_VOTE) - 1;
+
+  enum ProposalState {PrepayingGas, Started, Accepted, Rejected, InProbation, Completed}
 
   struct Proposal {
     address payable contributor;
     string title;
     string documentsLink;
     bytes documentsHash;
 
-    uint256 yesVotes;
-    uint256 noVotes;
+    ProposalState state;
+    uint256 probationEndDate;
+    uint256 probationPeriod;
+    uint256 approvalPercentage;
+
+    uint256 slots;
+    uint256 votesWords;
+    uint256 paidVotes;
+
+    mapping (uint256 => uint256) votes;
+    mapping (uint256 => uint256) voteCount;
+    mapping (uint256 => uint256) payGas1;
+    mapping (uint256 => uint256) payGas2;
+  }
+
+  mapping (uint256 => Proposal) public proposals;
+
+  function initTreasury() private {}
+
+  function msb(uint256 x) public pure returns (uint8 r) {
+    if (x >= 0x100000000000000000000000000000000) {x >>= 128; r += 128;}
+    if (x >= 0x10000000000000000) {x >>= 64; r += 64;}
+    if (x >= 0x100000000) {x >>= 32; r += 32;}
+    if (x >= 0x10000) {x >>= 16; r += 16;}
+    if (x >= 0x100) {x >>= 8; r += 8;}
+    if (x >= 0x10) {x >>= 4; r += 4;}
+    if (x >= 0x4) {x >>= 2; r += 2;}
+    if (x >= 0x2) r += 1; // No need to shift x anymore
+  }
+
+  function createProposal(
+      uint256 id,  // For test purposes until it's clear what will be used as id
+      address payable contributor,
+      string memory title,
+      string memory documents_link,
+      bytes memory documents_hash,
+      uint256 probation_period,
+      uint256 approval_percentage) public {
+    // require (?? proposals[id] == null)
+    proposals[id] = Proposal(contributor, title, documents_link, documents_hash, ProposalState.PrepayingGas,
```

Comment:
`Proposal storage p = proposals[id];`
it's better to set only the fields you care about using `p.field = value;` if we rearrange the fields, the code here as is could introduce a bug. It is also more readable, since the reader can see which fields is set to what.